File: lib/haveapi/action.rb
module HaveAPI
  class Action < Common
    obj_type :action
    has_attr :version
    has_attr :desc
    has_attr :route
    has_attr :resolve, ->(klass){ klass.respond_to?(:id) ? klass.id : nil }
    has_attr :http_method, :get
    has_attr :auth, true
    has_attr :aliases, []

    include Hookable

    has_hook :exec_exception

    attr_reader :message, :errors, :version

    class << self
      attr_reader :resource, :authorization, :examples

      def inherited(subclass)
        # puts "Action.inherited called #{subclass} from #{to_s}"

        subclass.instance_variable_set(:@obj_type, obj_type)

        if subclass.name
          # not an anonymouse class
          delayed_inherited(subclass)
        end
      end

      def delayed_inherited(subclass)
        resource = Kernel.const_get(subclass.to_s.deconstantize)

        inherit_attrs(subclass)
        inherit_attrs_from_resource(subclass, resource, [:auth])

        i = @input.clone
        i.action = subclass

        o = @output.clone
        o.action = subclass

        m = {}

        @meta.each do |k,v|
          m[k] = v && v.clone
          next unless v
          m[k].action = subclass
        end

        subclass.instance_variable_set(:@input, i)
        subclass.instance_variable_set(:@output, o)
        subclass.instance_variable_set(:@meta, m)

        begin
          subclass.instance_variable_set(:@resource, resource)
          subclass.instance_variable_set(:@model, resource.model)
        rescue NoMethodError
          return
        end
      end

      def initialized?
        @initialize
      end

      def initialized
        @initialize = true
      end

      def model_adapter(layout)
        ModelAdapter.for(layout, resource.model)
      end

      def input(layout = nil, namespace: nil, &block)
        if block
          @input ||= Params.new(:input, self)
          @input.layout = layout
          @input.namespace = namespace
          @input.instance_eval(&block)

          model_adapter(@input.layout).load_validators(model, @input) if model
        else
          @input
        end
      end

      def output(layout = nil, namespace: nil, &block)
        if block
          @output ||= Params.new(:output, self)
          @output.layout = layout
          @output.namespace = namespace
          @output.instance_eval(&block)
        else
          @output
        end
      end

      def meta(type = :object, &block)
        if block
          @meta ||= {object: nil, global: nil}
          @meta[type] ||= Metadata::ActionMetadata.new
          @meta[type].action = self
          @meta[type].instance_exec(&block)
        else
          @meta[type]
        end
      end

      def authorize(&block)
        @authorization = Authorization.new(&block)
      end

      def example(title = '', &block)
        @examples ||= []
        e = Example.new(title)
        e.instance_eval(&block)
        @examples << e
      end

      def build_route(prefix)
        route = @route || to_s.demodulize.underscore

        if !route.is_a?(String) && route.respond_to?(:call)
          route = route.call(self.resource)
        end

        prefix + route % {resource: self.resource.to_s.demodulize.underscore}
      end

      def describe(context)
        authorization = (@authorization && @authorization.clone) || Authorization.new

        return false if (context.endpoint || context.current_user) && !authorization.authorized?(context.current_user)

        route_method = context.action.http_method.to_s.upcase
        context.authorization = authorization

        if context.endpoint
          context.action_instance = context.action.from_context(context)
          context.action_prepare = context.action_instance.prepare
        end

        {
            auth: @auth,
            description: @desc,
            aliases: @aliases,
            input: @input ? @input.describe(context) : {parameters: {}},
            output: @output ? @output.describe(context) : {parameters: {}},
            meta: @meta ? @meta.merge(@meta) { |_, v| v && v.describe(context) } : nil,
            examples: @examples ? @examples.map { |e| e.describe } : [],
            url: context.resolved_url,
            method: route_method,
            help: "#{context.url}?method=#{route_method}"
        }
      end

      # Inherit attributes from resource action is defined in.
      def inherit_attrs_from_resource(action, r, attrs)
        begin
          return unless r.obj_type == :resource

        rescue NoMethodError
          return
        end

        attrs.each do |attr|
          action.method(attr).call(r.method(attr).call)
        end
      end

      def from_context(c)
        ret = new(nil, c.version, c.params, nil, c)
        ret.instance_exec do
          @safe_params = @params.dup
          @authorization = c.authorization
        end

        ret
      end
    end

    def initialize(request, version, params, body, context)
      @request = request
      @version = version
      @params = params
      @params.update(body) if body
      @context = context
      @context.action = self.class
      @context.action_instance = self
      @reply_meta = {object: {}, global: {}}

      class_auth = self.class.authorization

      if class_auth
        @authorization = class_auth.clone
      else
        @authorization = Authorization.new {}
      end
    end

    def validate!
      begin
        @params = validate
      rescue ValidationError => e
        error(e.message, e.to_hash)
      end
    end

    def authorized?(user)
      @current_user = user
      @authorization.authorized?(user)
    end

    def current_user
      @current_user
    end

    def params
      @safe_params
    end

    def input
      @safe_params[ self.class.input.namespace ] if self.class.input
    end

    def request
      @request
    end

    def meta
      @metadata
    end

    def set_meta(hash)
      @reply_meta[:global].update(hash)
    end

    # Prepare object, set instance variables from URL parameters.
    # This method should return queried object. If the method is
    # not implemented or returns nil, action description will not
    # contain link to an associated resource.
    # --
    # FIXME: is this correct behaviour?
    # ++
    def prepare

    end

    def pre_exec

    end

    # This method must be reimplemented in every action.
    # It must not be invoked directly, only via safe_exec, which restricts output.
    def exec
      ['not implemented']
    end

    # Calls exec while catching all exceptions and restricting output only
    # to what user can see.
    # Return array +[status, data|error, errors]+
    def safe_exec
      ret = catch(:return) do
        begin
          validate!
          prepare
          pre_exec
          exec
        rescue Exception => e
          tmp = call_class_hooks_as_for(Action, :exec_exception, args: [self, e])

          if tmp.empty?
            p e.message
            puts e.backtrace
            error('Server error occurred')
          end

          error(tmp[:message]) unless tmp[:status]
        end
      end

      if ret
        output = self.class.output

        if output
          safe_ret = nil
          adapter = self.class.model_adapter(output.layout)

          case output.layout
            when :object
              out = adapter.output(@context, ret)
              safe_ret = @authorization.filter_output(
                  self.class.output.params,
                  out
              )
              @reply_meta[:global].update(out.meta)

            when :object_list
              safe_ret = []

              ret.each do |obj|
                out = adapter.output(@context, obj)

                safe_ret << @authorization.filter_output(
                    self.class.output.params,
                    out
                )
                safe_ret.last.update({Metadata.namespace => out.meta}) unless meta[:no]
              end

            when :hash
              safe_ret = @authorization.filter_output(
                          self.class.output.params,
                          adapter.output(@context, ret))

            when :hash_list
              safe_ret = ret
              safe_ret.map! do |hash|
                @authorization.filter_output(
                    self.class.output.params,
                    adapter.output(@context, hash))
              end

            else
              safe_ret = ret
          end

          ns = {output.namespace => safe_ret}
          ns[Metadata.namespace] = @reply_meta[:global] unless meta[:no]

          [true, ns]

        else
          [true, {}]
        end

      else
        [false, @message, @errors]
      end
    end

    def v?(v)
      @version == v
    end

    input {}
    output {}
    meta(:global) do
      input do
        bool :no, label: 'Disable metadata'
      end
    end

    protected
    def with_restricted(*args)
      if args.empty?
        @authorization.restrictions
      else
        args.first.update(@authorization.restrictions)
      end
    end

    # Convert parameter names to corresponding DB names.
    # By default, input parameters are used for the translation.
    def to_db_names(hash, src=:input)
      return {} unless hash

      params = self.class.method(src).call.params
      ret = {}

      hash.each do |k, v|
        k = k.to_sym
        hit = false

        params.each do |p|
          if k == p.name
            ret[p.db_name] = v
            hit = true
            break
          end
        end

        ret[k] = v unless hit
      end

      ret
    end

    # Convert DB names to corresponding parameter names.
    # By default, output parameters are used for the translation.
    def to_param_names(hash, src=:output)
      return {} unless hash

      params = self.class.method(src).call.params
      ret = {}

      hash.each do |k, v|
        k = k.to_sym
        hit = false

        params.each do |p|
          if k == p.db_name
            ret[p.name] = v
            hit = true
            break
          end
        end

        ret[k] = v unless hit
      end

      ret
    end

    def ok(ret={})
      throw(:return, ret)
    end

    def error(msg, errs={})
      @message = msg
      @errors = errs
      throw(:return, false)
    end

    private
    def validate
      # Validate standard input
      @safe_params = @params.dup
      input = self.class.input

      if input
        # First check layout
        input.check_layout(@safe_params)

        # Then filter allowed params
        case input.layout
          when :object_list, :hash_list
            @safe_params[input.namespace].map! do |obj|
              @authorization.filter_input(
                  self.class.input.params,
                  self.class.model_adapter(self.class.input.layout).input(obj))
            end

          else
            @safe_params[input.namespace] = @authorization.filter_input(
                self.class.input.params,
                self.class.model_adapter(self.class.input.layout).input(@safe_params[input.namespace]))
        end

        # Remove duplicit key
        @safe_params.delete(input.namespace.to_s)

        # Now check required params, convert types and set defaults
        input.validate(@safe_params)
      end

      # Validate metadata input
      auth = Authorization.new { allow }
      @metadata = {}

      return if input && %i(object_list hash_list).include?(input.layout)

      [:object, :global].each do |v|
        meta = self.class.meta(v)
        next unless meta

        raw_meta = nil

        [Metadata.namespace, Metadata.namespace.to_s].each do |ns|
          params = v == :object ? (@params[input.namespace] && @params[input.namespace][ns]) : @params[ns]
          next unless params

          raw_meta = auth.filter_input(
              meta.input.params,
              self.class.model_adapter(meta.input.layout).input(params)
          )

          break if raw_meta
        end

        next unless raw_meta

        @metadata.update(meta.input.validate(raw_meta))
      end
    end
  end
end

File: lib/haveapi/action.ex
defmodule HaveAPI.Action do
  alias HaveAPI.Action.Input
  alias HaveAPI.Action.Output

  defmacro __using__(_opts) do
    quote location: :keep do
      @haveapi_parent nil
      unquote(HaveAPI.Action.setup)

      defmacro __using__(_opts) do
        quote do
          parent = unquote(__MODULE__)

          @haveapi_parent parent
          unquote(HaveAPI.Action.setup)

          Enum.each(
            [:method, :route, :aliases, :auth],
            fn v ->
              Module.put_attribute(
                __MODULE__,
                :"haveapi_#{v}",
                apply(parent, v, [])
              )
            end
          )

          Enum.each(
            [:input, :output],
            fn dir ->
              params = parent.params(dir)

              if params do
                Module.put_attribute(
                  __MODULE__,
                  :"haveapi_parent_#{dir}_layout",
                  parent.layout(dir)
                )
                Module.put_attribute(
                  __MODULE__,
                  :"haveapi_parent_#{dir}",
                  params
                )
              end
            end
          )

          unquote(__MODULE__.use_template)
        end
      end
    end
  end

  defmacro method(v) do
    quote do: @haveapi_method unquote(v)
  end

  defmacro route(v) do
    quote do: @haveapi_route unquote(v)
  end

  defmacro desc(v) do
    quote do: @haveapi_desc unquote(v)
  end

  defmacro aliases(v) do
    quote do: @haveapi_aliases (@haveapi_aliases ++ unquote(v))
  end

  defmacro auth(v) when is_boolean(v) do
    quote do: @haveapi_auth unquote(v)
  end

  defmacro input(layout \\ nil, [do: block]) do
    quote do
      @haveapi_input true

      layout = case unquote(layout) do
        nil ->
          @haveapi_parent_input_layout || :hash

        any ->
          any
      end

      parent_in = @haveapi_parent_input

      defmodule Input do
        use HaveAPI.Parameters.Dsl, layout: layout

        Enum.each(
          parent_in,
          fn v -> @haveapi_params [v | @haveapi_params] end
        )

        unquote(block)
      end
    end
  end

  defmacro output(layout \\ nil, [do: block]) do
    quote do
      @haveapi_output true

      layout = case unquote(layout) do
        nil ->
          @haveapi_parent_output_layout || :hash

        any ->
          any
      end

      parent_out = @haveapi_parent_output

      defmodule Output do
        use HaveAPI.Parameters.Dsl, layout: layout

        Enum.each(
          parent_out,
          fn v -> @haveapi_params [v | @haveapi_params] end
        )

        unquote(block)
      end
    end
  end

  defmacro meta(:local, [do: block]) do
    quote do
      @haveapi_meta_local true
      parent = @haveapi_parent

      defmodule LocalMeta do
        use HaveAPI.Meta

        if parent && parent.has_meta?(:local) do
          Enum.each([:input, :output], fn dir ->
            attr = :"haveapi_parent_#{dir}"
            Module.register_attribute(__MODULE__, attr, [])

            params = parent.meta(:local).params(dir)

            if params do
              Module.put_attribute(__MODULE__, attr, params)
            end
          end)
        end

        unquote(block)
      end
    end
  end

  defmacro meta(:global, [do: block]) do
    quote do
      @haveapi_meta_global true
      parent = @haveapi_parent

      defmodule GlobalMeta do
        use HaveAPI.Meta

        if parent && parent.has_meta?(:global) do
          Enum.each([:input, :output], fn dir ->
            attr = :"haveapi_parent_#{dir}"
            Module.register_attribute(__MODULE__, :attr, [])

            params = parent.meta(:global).params(dir)

            if params do
              Module.put_attribute(__MODULE__, attr, params)
            end
          end)
        end

        unquote(block)
      end
    end
  end

  defmacro __before_compile__(_env) do
    quote do
      if @haveapi_parent_input_layout && !@haveapi_input do
        input do: nil
      end

      if @haveapi_parent_output_layout && !@haveapi_output do
        output do: nil
      end

      parent = @haveapi_parent
      if parent do
        if parent.has_meta?(:local) && !@haveapi_meta_local do
          meta(:local) do
            if parent.meta(:local).params(:input) do
              input do: nil
            end

            if parent.meta(:local).params(:output) do
              output do: nil
            end
          end
        end

        if parent.has_meta?(:global) && !@haveapi_meta_global do
          meta(:global) do
            if parent.meta(:global).params(:input) do
              input do: nil
            end

            if parent.meta(:global).params(:output) do
              output do: nil
            end
          end
        end
      end

      def method, do: @haveapi_method

      def route, do: @haveapi_route

      def resolve_route(route, [resource]) do
        route = Regex.replace(~r/%{resource}/, route, resource.name, global: true)
        route = Regex.replace(~r/%{action}/, route, name(), global: true)
      end

      def resolve_route(route, [first, second]) do
        route = Regex.replace(~r/%{resource}/, route, first.name, global: false)
        route = Regex.replace(~r/%{resource}/, route, second.name, global: true)
        route = Regex.replace(~r/%{action}/, route, name(), global: true)
      end

      def name do
        Module.split(__MODULE__) |> List.last |> String.downcase
      end

      def desc, do: @haveapi_desc

      def aliases, do: @haveapi_aliases

      def auth, do: @haveapi_auth

      def io(:input) do
        Module.concat(__MODULE__, :Input)
      end

      def io(:output) do
        Module.concat(__MODULE__, :Output)
      end

      def layout(dir) do
        apply(io(dir), :layout, [])

      rescue
        UndefinedFunctionError -> nil
      end

      def params(dir) do
        apply(io(dir), :params, [])

      rescue
        UndefinedFunctionError -> nil
      end

      def has_meta?(:local), do: @haveapi_meta_local
      def has_meta?(:global), do: @haveapi_meta_global

      def meta(:local), do: Module.concat(__MODULE__, :LocalMeta)
      def meta(:global), do: Module.concat(__MODULE__, :GlobalMeta)

      def template, do: @haveapi_parent
    end
  end

  def setup do
    quote do
      @haveapi_method :get
      @haveapi_route ""
      @haveapi_desc ""
      @haveapi_aliases []
      @haveapi_auth true
      @haveapi_input false
      @haveapi_parent_input_layout nil
      @haveapi_parent_input []
      @haveapi_output false
      @haveapi_parent_output_layout nil
      @haveapi_parent_output []
      @haveapi_meta_local false
      @haveapi_meta_global false
      @before_compile HaveAPI.Action

      import HaveAPI.Action

      def authorize(_req, _user), do: if auth(), do: :deny, else: :allow

      def use_template, do: nil

      defoverridable [authorize: 2, use_template: 0]
    end
  end

  def execute(ctx, conn) do
    ctx = %{ctx | conn: conn, user: HaveAPI.Authentication.user(conn)}
    req = %HaveAPI.Request{context: ctx, conn: conn, user: ctx.user}
    res = %HaveAPI.Response{context: ctx, conn: conn}

    with true <- authenticated?(ctx.action.auth, ctx.user),
         req <- Input.fetch_path_parameters(req),
         {:ok, req} <- Input.fetch_parameters(req),
         {:ok, req, params} <- HaveAPI.Authorization.authorize(req),
         :ok <- HaveAPI.Validator.validate(req, params),
         data <- do_exec(req),
         output = ctx.action.layout(:output),
         res <- Output.build(data, output, res),
         {:ok, res} <- HaveAPI.Authorization.authorize(res),
         {:ok, res} <- Output.filter(res, output),
         {:ok, res} <- Output.filter_meta(res) do
      reply(res)
    else
      {:error, msg} when is_binary(msg) ->
        reply(%{res | status: false, message: msg})

      {:error, msg, opts} when is_list(opts) ->
        reply(%{res | status: false, message: msg} |> Map.merge(Map.new(opts)))
    end
  end

  def internal(%HaveAPI.Context{conn: conn} = ctx, opts \\ []) do
    req = %HaveAPI.Request{
      context: ctx,
      conn: conn,
      user: ctx.user,
      params: opts[:params] && map_path_params(ctx, opts[:params]),
      input: opts[:input],
    }

    with {:ok, req, params} <- HaveAPI.Authorization.authorize(req),
         :ok <- HaveAPI.Validator.validate(req, params),
         data <- do_exec(req),
         res = %HaveAPI.Response{context: ctx, conn: conn},
         output = ctx.action.layout(:output),
         res <- Output.build(data, output, res),
         {:ok, res} <- HaveAPI.Authorization.authorize(res),
         {:ok, res} <- Output.filter(res, output),
    do: res
  end

  defp authenticated?(true, user) do
    if user, do: true, else: {:error, "Access forbidden", http_status: 403}
  end

  defp authenticated?(false, _user), do: true

  defp map_path_params(ctx, params) do
    route = Path.join(
      [ctx.prefix] ++
      Enum.map(ctx.resource_path, &(&1.route)) ++
      [ctx.action.route]
    ) |> ctx.action.resolve_route(ctx.resource_path)

    path_params = Regex.scan(~r{:([^/]+)}, route)
      |> Enum.map(fn [_m, v] -> String.to_atom(v) end)

    Enum.zip(path_params, params)
  end

  defp do_exec(req) do
    apply(req.context.action, :exec, [req])
  end

  defp reply(res), do: HaveAPI.Protocol.send_data(res)
end
